cookbooks/creating_a_behat_extension.rst

```diff
@@ -0,0 +1,278 @@
+
+To pass configuration values to our ``HelloWorldContext``, we need to create an initializer.
+
+.. code-block::
+  src/
+      Context/
+          Initializer/
+              HelloWorldInitializer.php   # This will handle context initialization
+            HelloWorldContext.php
+      ServiceContainer/
+        HelloWorldExtension.php
+
+The code for ``HelloWorldInitializer.php``:
+
+.. code-block:: php
+
+  <?php
+
+  namespace HelloWorld\Context\Initializer;
+
+  use HelloWorld\Context\HelloWorldContext;
+  use Behat\Behat\Context\Context;
+  use Behat\Behat\Context\Initializer\ContextInitializer;
+
+  class HelloWorldInitializer implements ContextInitializer
+  {
+      private string $text;
+      private bool $enable;
+
+      public function __construct(string $text, bool $enable)
+      {
+          $this->text = $text;
+          $this->enable = $enable;
+      }
+
+      public function initializeContext(Context $context)
+      {
+          /*
+           * At the start of every scenario, behat will create a new instance of every `Context`
+           * registered in your project. It will then call this method with each new `Context` in
+           * turn. If you want to initialise multiple contexts, you can of course give them an
+           * interface and check for that here.
+           */
+          if (!$context instanceof HelloWorldContext) {
+              return;
+          }
+
+          $context->initializeConfig($this->enable, $this->text);
+      }
+  }
+
+We need to register the initializer definition within the Behat container through the ``HelloWorldExtension``, ensuring it gets loaded:
+
+.. code-block:: php
+
+  <?php
+
+  // ...
+
+  use Symfony\Component\DependencyInjection\Definition;
+  use Behat\Behat\Context\ServiceContainer\ContextExtension;
+
+  class HelloWorldExtension implements Extension
+  {
+      // ...
+
+      public function load(ContainerBuilder $container, array $config)
+      {
+          $definition = new Definition(HelloWorldInitializer::class, [
+              $config['text'],
+              $config['enable'],
+          ]);
+          $definition->addTag(ContextExtension::INITIALIZER_TAG);
+          $container->setDefinition('helloworld_extension.context_initializer', $definition);
+      }
+
+      // ...
+  }
+
+To complete the extension, we must add methods to ``HelloWorldContext`` to receive the configuration values and use those in the hooks:
+
+.. code-block:: php
+
+  // ...
+
+  class HelloWorldContext implements Context
+  {
+      private bool $enable = false;
+      private string $text;
+
+      public function initializeConfig(bool $enable, string $text)
+      {
+          $this->enable = $enable;
+          $this->text = $text;
+      }
+
+      /** @Given I say Hello World */
+      public function helloWorld()
+      {
+          if ($this->enable === false) {
+              return;
+          }
+  
+          throw new PendingException($this->text);
+      }
+  }
```

Comment:
I don't think this code block is functionally different to the first one (just the `helloWorld` method is formatted/structured slightly different. Did you mean to leave out the `initializeConfig` method in the first example and then add it here? 

I think the example is simple enough now that you could probably just include the whole `HelloWorldContext` once at the beginning and explain that the main job of the extension is to call `initializeConfig` with the values? And then leave this section out?

Reply:
It came from a copy / paste from a test repo, and forgot to remove the method.

But you're right, it simple enough now to have a single block at the beginning.
It's done 👌